eth2util/enr/enr.go

```diff
@@ -0,0 +1,158 @@
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package enr provides a minimal implementation of Ethereum Node Records (ENR).
+package enr
+
+import (
+	"crypto/ecdsa"
+	"encoding/base64"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/obolnetwork/charon/app/errors"
+	"github.com/obolnetwork/charon/eth2util/rlp"
+)
+
+const (
+	// keySecp256k1 is the key used to store the secp256k1 public key in the record.
+	keySecp256k1 = "secp256k1"
+
+	// keyID is the key used to store the identity scheme in the record, only v4 supported.
+	keyID = "id"
+	valID = "v4"
+)
+
+// Parse parses the given base64 encoded string into a record.
+func Parse(enrStr string) (Record, error) {
+	if !strings.HasPrefix(enrStr, "enr:") {
+		return Record{}, errors.New("missing 'enr:' prefix")
+	}
+
+	raw, err := base64.RawURLEncoding.DecodeString(enrStr[4:])
+	if err != nil {
+		return Record{}, errors.Wrap(err, "invalid base64 encoding")
+	}
+
+	elements, err := rlp.DecodeBytesList(raw)
+	if err != nil {
+		return Record{}, errors.Wrap(err, "invalid rlp encoding")
+	}
+
+	if len(elements) < 4 {
+		return Record{}, errors.New("invalid enr record, too few elements")
+	}
+	if len(elements)%2 != 0 {
+		return Record{}, errors.New("invalid enr record, odd number of elements")
+	}
+
+	r := Record{
+		Signature: elements[0],
+	}
+
+	for i := 2; i < len(elements); i += 2 {
+		switch string(elements[i]) {
+		case keySecp256k1:
+			r.PubKey, err = crypto.DecompressPubkey(elements[i+1])
+			if err != nil {
+				return Record{}, errors.Wrap(err, "invalid secp256k1 public key")
+			}
+		case keyID:
+			if string(elements[i+1]) != valID {
+				return Record{}, errors.New("non-v4 identity scheme not supported")
+			}
+		}
+	}
+
+	if r.PubKey == nil {
+		return Record{}, errors.New("missing secp256k1 public key")
+	}
+
+	if err := verify(r.PubKey, r.Signature, rlp.EncodeBytesList(elements[1:])); err != nil {
+		return Record{}, err
+	}
+
+	return r, nil
+}
+
+// New returns a new enr record for the given private key.
+func New(privkey *ecdsa.PrivateKey) (Record, error) {
+	sig, err := sign(privkey)
+	if err != nil {
+		return Record{}, err
+	}
+
+	return Record{
+		PubKey:    &privkey.PublicKey,
+		Signature: sig,
+	}, nil
+}
+
+// Record represents an Ethereum Node Record.
+type Record struct {
```

Comment:
is this compatible with records already generated by enr.Record from geth?

Reply:
no